fe/fe-core/src/main/java/com/starrocks/sql/common/ExpressionPartitionUtil.java

```diff
@@ -1,205 +1,224 @@
+                                                        Map<String, Range<PartitionKey>> mvRangeMap,
+                                                        String granularity) {
+        Map<String, Range<PartitionKey>> rollupRange = mappingRangeList(baseRangeMap, granularity);
+        Map<String, Range<PartitionKey>> adds = diffRange(rollupRange, mvRangeMap);
+        Map<String, Range<PartitionKey>> deletes = diffRange(mvRangeMap, rollupRange);
+        return new PartitionDiff(adds, deletes);
+
     }
 
-    public static Range<PartitionKey> getPartitionKeyRange(
-            Expr expr, Column partitionColumn,
-            Collection<Range<PartitionKey>> existPartitionKeyRanges,
-            Range<PartitionKey> basePartitionRange,
-            int basePartitionRangeIndex) {
-        LiteralExpr baseLowerLiteralExpr = basePartitionRange.lowerEndpoint().getKeys().get(basePartitionRangeIndex);
-        LiteralExpr baseUpperLiteralExpr = basePartitionRange.upperEndpoint().getKeys().get(basePartitionRangeIndex);
-        if (expr instanceof SlotRef) {
-            try {
-                PartitionKey lowerBound = PartitionKey.createPartitionKey(
-                        Collections.singletonList(new PartitionValue(baseLowerLiteralExpr.getStringValue())),
-                        Collections.singletonList(partitionColumn));
-                PartitionKey upperBound = PartitionKey.createPartitionKey(
-                        Collections.singletonList(new PartitionValue(baseUpperLiteralExpr.getStringValue())),
-                        Collections.singletonList(partitionColumn));
-                return Range.closedOpen(lowerBound, upperBound);
-            } catch (AnalysisException e) {
-                throw new SemanticException("Create Partition Range failed:" + e.getMessage());
-            }
-        } else if (expr instanceof FunctionCallExpr) {
-            FunctionCallExpr functionCallExpr = (FunctionCallExpr) expr;
-
-            LiteralExpr lowerLiteralExpr = ExpressionPartitionUtil.getBound(
-                    functionCallExpr, baseLowerLiteralExpr, partitionColumn.getPrimitiveType(), 0);
-            LiteralExpr upperLiteralExpr = ExpressionPartitionUtil.getBound(
-                    functionCallExpr, baseUpperLiteralExpr, partitionColumn.getPrimitiveType(), 0);
-            if (ExpressionPartitionUtil.compareBound(baseUpperLiteralExpr, upperLiteralExpr) > 0) {
-                // increase upperBound
-                upperLiteralExpr = ExpressionPartitionUtil.getBound(
-                        functionCallExpr, upperLiteralExpr, partitionColumn.getPrimitiveType(), 1);
-            }
+    public static Map<String, Range<PartitionKey>> mappingRangeList(Map<String, Range<PartitionKey>> baseRangeMap,
+                                                              String granularity) {
+        Set<LocalDateTime> timePointSet = Sets.newTreeSet();
```

Comment:
Should we be limited to time types ?

Reply:
I think DateTime for rollup will be enough

Reply:
you are right. we should also support type DATE